Support setting typing status

PIG208 opened the pull request:
Follow-up to and stacked on top of #888

Fixes: #666

PIG208 commented:
Switched to a different implementation for the compose box updates.

chrisbobbe reviewed (commented):
Thanks! Comments below.

PIG208 commented on lib/model/typing_status.dart:
```diff
@@ -84,3 +87,138 @@ class TypingStatus extends ChangeNotifier {
+  ///
+  /// See docs/subsystems/typing-indicators.md for detailed background on the
+  /// typing indicators system.
+  void handleTypingStatusUpdate({SendableNarrow? destination}) {
```
I think this looks like a good refactor. The `null` = "stop typing" is less obvious.

gnprice replied:
Yeah, splitting these up sounds like a good API improvement.

PIG208 commented on lib/model/store.dart:
```diff
@@ -364,6 +372,7 @@ class PerAccountStore extends ChangeNotifier with EmojiStore, ChannelStore, Mess
   final Map<int, User> users;
 
   final TypingStatus typingStatus;
+  final TypingNotifier typingNotifier;
```
Yeah, `TypingNotifier` itself only maintains its internal states, so I wasn't sure if `PerAccountStore` is the best fit. Moving it to `UpdateMachine` sounds like a creative idea, but I'm not sure do we typically want to access it from our widget code.

gnprice replied:
I think of `UpdateMachine` as mainly having the job of keeping its associated `PerAccountStore` up to date with data from the server.

Then with `registerNotificationToken` it got a bit of mission creep. What really ties that method to the rest of `UpdateMachine` is: it's code that in the live app is associated one-to-one with a `PerAccountStore`, but that on its own initiative goes and makes HTTP requests to the server, and that therefore we don't want to be active in our tests except where specifically relevant. (Because if it were active, it'd be a pain to manage those HTTP requests interleaving with whatever other HTTP requests any particular test is interested in.) There's probably a better way to organize that but that's how it currently is.

With `TypingNotifier`, there's no need to take any particular steps to suppress it in the bulk of our tests — it only does anything when its methods get called. The more narrowly-targeted, and explicit, suppression that a handful of files end up needing in this PR works fine — and the explicitness is nice. So `PerAccountStore` is a good home.

PIG208 commented on lib/widgets/compose_box.dart:
```diff
@@ -272,12 +272,14 @@ class ComposeContentController extends ComposeController<ContentValidationError>
   });
 
   final Narrow narrow;
+  final SendableNarrow Function() getTypingNotificationNarrow;
```
Yeah, editing topic input always blurs content input, which sends a "typing stopped" notice.

gnprice commented on lib/widgets/compose_box.dart:
```diff
@@ -287,6 +289,61 @@ class _ContentInput extends StatefulWidget {
+  }
+
+  void _contentChanged() {
+    if (_prevText == widget.controller.text) {
```
You can find a version(*) of `String.==` if you have your IDE navigate to `_StringBase` (it's in `bin/cache/pkg/sky_engine/lib/_internal/vm/lib/string_patch.dart` in your Flutter tree).

```dart
  @pragma("vm:exact-result-type", bool)
  bool operator ==(Object other) {
    if (identical(this, other)) {
      return true;
    }
    if (other is String && this.length == other.length) {
      final len = this.length;
      for (int i = 0; i < len; i++) {
        if (this.codeUnitAt(i) != other.codeUnitAt(i)) {
          return false;
        }
      }
      return true;
    }
    return false;
  }
```

As you can see there, it has a fast path if the strings are identical, and another if they're different lengths, returning true and false respectively. So the linear-time comparison happens only if the new string is a different instance yet has the same length.

I think this is probably fine. If we discover it's a performance issue, the fix might be to only check `identical`, and if not identical then assume the content was actually edited.

(*) I'm not sure that's the actual version that gets run, because the two concrete implementations `_OneByteString` and `_TwoByteString` override it like so:
```dart
  @pragma("vm:recognized", "asm-intrinsic")
  @pragma("vm:exact-result-type", bool)
  // Intrinsic is more efficient than an inlined body even for the small
  // strings.
  @pragma('vm:never-inline')
  bool operator ==(Object other) {
    return super == other;
  }
```

The `asm-intrinsic` thing sounds like the Dart VM may replace calls to that method with calls to an implementation written in assembly. I'd expect the assembly implementation to have all the same algorithmic optimizations as the pure-Dart version, though — it'll just have better constant factors.

chrisbobbe replied:
Oh OK, interesting!

> So the linear-time comparison happens only if the new string is a different instance yet has the same length.

Yeah, this does seem rare here.

chrisbobbe commented:
LGTM, but how about also calling `stoppedComposing` at/near the top of the `_send` helper? I think we already send "typing stopped" notices when the send button is tapped, but kind of incidentally, because we clear the content input—

```dart
    widget.contentController.clear();
```

—and our content-input listener responds to that by calling `stoppedComposing`. I think an explicit `stoppedComposing` call would be helpful, especially because we have a TODO for delaying the clear-input action:

```dart
    widget.contentController.clear();

    try {
      // TODO(#720) clear content input only on success response;
      //   while waiting, put input(s) and send button into a disabled
      //   "working on it" state (letting input text be selected for copying).
```

and we want to send the notice as soon as it applies, i.e., as soon as the send button is tapped.

(As a consequence, I guess we'll end up making a redundant `stoppedComposing` call, but that should be harmless.)

chrisbobbe commented:
Marking for Greg's review. Thanks for the revision!

PIG208 commented:
Thanks for the review! I wasn't clear about the status of the PR because I didn't post a comment after the pushes yesterday. Just worked on some improvements and addressed the new feedback!

gnprice reviewed (commented):
Thanks @PIG208, and thanks @chrisbobbe for the previous reviews!

I've read through the first two commits:
536566949 fix link; binding: Add stopwatch.
c40ec190f model: Add TypingNotifier

Generally it all looks good; various comments below.

PIG208 commented on lib/model/typing_status.dart:
```diff
@@ -84,3 +87,154 @@ class TypingStatus extends ChangeNotifier {
+  }
+
+  void _maybePingServer() {
+    if (_timeSinceLastPing == null
+      || _timeSinceLastPing!.elapsed > typingStartedWaitPeriod) {
```
Right. It can't because we already checked `_currentDestination != null`.

PIG208 commented on test/model/typing_status_test.dart:
```diff
@@ -210,4 +248,247 @@ void main() {
+
+      async.elapse(waitTime);
+      // t = 100: The idle timer is reset to typingStoppedWaitPeriod.
+      sendTypingNotification(async, narrow);
```
This request [doesn't get sent](https://github.com/zulip/zulip-flutter/pull/934#discussion_r1796303132) until the `async.elapse` call after `model.keystroke`, because we `unawaited` the `setTypingStatus` call. We should probably reintroduce futures to `keystroke` et al. so that the callers can `await` them.

gnprice replied:
Hmm. That linked discussion is in a pretty different context: it was a widget test, which manages async differently from `awaitFakeAsync`. And in that situation, returning a future and trying to await it is precisely the thing that *doesn't* work at all, and instead gets stuck.

As described in my other couple of comments just now, it looks like we don't want the callers in the actual app to await this. Which makes sense because we really don't care about the results of this API request — there's nothing we'll do with the response, and if it fails then we'll just shrug because it's inherently transient information anyway.

Given that the actual app doesn't want to await this, there is still an option to return a future just for the sake of tests, but it'd be better to avoid that if we can. In general, it's best to where possible keep any given piece of internal API the way the consumers in the actual app want it, and not complicate it for the sake of tests except where necessary.

So if an `async.elapse()` call is what it takes to make the tests work fine with the simpler API, then let's go for that.

(I'm still a bit puzzled that even that much is necessary, because it looks like the implementation goes ahead and *makes* the request synchronously — the only thing that happens asynchronously is for the request to finish.)

PIG208 replied:
I found that what actually happens is that the `async.elapse` is only needed when we check the timers in some cases. Both `new Timer` and `setTypingStatus` creates a timer tracked by `async.pendingTimers`, so whenever we want to see if `new Timer` works, or that `stoppedComposing` cancels the timer, there will also be a pending request.

gnprice replied:
I don't totally follow. What's an example of when we want to see if `new Timer` works? (Surely it always works, right? And if it somehow doesn't, that's an upstream problem.)

PIG208 replied:
Yeah, `new Timer` itself works. We want to test that `keystroke` results in a timer being created (or extended), and that `stoppedComposing` cancels the timer. Both methods might send an API request that also counts toward `async.pendingTimers` with `Duration.zero`.

PIG208 commented:
Updated the PR. Thanks for the review!

gnprice commented on test/model/typing_status_test.dart:
```diff
@@ -210,4 +248,247 @@ void main() {
+      // [waitInterval] should not be long enough
+      // to trigger a "typing stopped" notice.
+      assert(waitInterval < model.typingStoppedWaitPeriod);
+      // [waitInterval] should not be long enough either,
+      // that we only send the "typing started" notice once.
+      assert(waitInterval < model.typingStartedWaitPeriod);
```
```suggestion
      // [waitInterval] should be short enough
      // that the loop below runs more than once.
      assert(waitInterval < model.typingStartedWaitPeriod);
```

I read 'send the "typing started" notice' as meaning actually making the request, not just calling `model.keystroke`.

gnprice reviewed (commented):
OK, reread the first two commits:

11122dd47 fix link; binding: Add stopwatch.
ffeca32bd model: Add TypingNotifier

Comments here and above.

Those commits are quite close now. I think the revision will make some changes to the later commits, though, so I'll leave this review here for now and switch to reviewing other PRs.

PIG208 commented:
Thanks. I restored the simpler synchronous API and left a comment on the `async.elpase` discussion. Should be ready for review again!

gnprice commented on lib/widgets/compose_box.dart:
```diff
@@ -269,19 +269,99 @@ class ComposeContentController extends ComposeController<ContentValidationError>
+    final store = PerAccountStoreWidget.of(context);
+    (widget.controller.text.isEmpty)
+      ? store.typingNotifier.stoppedComposing()
+      : store.typingNotifier.keystroke(widget.destination);
```
Rereading the API docs:
https://zulip.com/api/set-typing-status

the spec for what a client should do includes:
> While the user continues to actively type or otherwise interact with the compose UI (e.g. interacting with the compose box emoji picker), send regular `"op": "start"` requests to this endpoint

Thinking about that, here's a couple more situations where we should be sending such requests:
* The user hits one of the compose buttons. (Eventually if they do choose an image, etc., then that will cause a typing notice anyway because it'll edit the text in the input; but that may be many seconds later. In a fast-moving conversation, it may make a difference to let other participants know immediately that they're preparing something to send.)
* The text input's selection changes, as opposed to its text. That's what it'll look like to us if the user is moving their cursor around the input, preparing their next edit; or if they're selecting a part of their draft to cut and perhaps move around.
* Ideally, also if the user scrolls around the results of an autocomplete. If this one is at all complicated to implement, though, it's low priority — I think generally people are not likely to do this for more than a couple of seconds, between typing more letters to narrow down the autocomplete and then either picking a result or giving up.

Of those, the second one can be squashed into the main commit here, since it's just removing some logic that that one adds; the other two should probably be separate commits, just like your existing followup commit for app lifecycle changes.

gnprice commented on test/widgets/compose_box_test.dart:
```diff
@@ -195,10 +206,163 @@ void main() {
+      unawaited(navigator.push(MaterialAccountWidgetRoute(
+        accountId: account.id, page: const ComposeBox(narrow: narrow))));
+      await tester.pumpAndSettle();
+
+      store = await testBinding.globalStore.perAccount(account.id);
+      connection = store.connection as FakeApiConnection;
```
nit: put these lines before the widget operations — that just makes this more similar to `prepareComposeBox`, making it easier to compare and see the remaining differences

gnprice commented on test/widgets/compose_box_test.dart:
```diff
@@ -195,10 +206,163 @@ void main() {
+    });
+
+    testWidgets('for content input, unfocusing sends "stopped typing" notice '
+                'and refocusing does not send "started typing notice', (tester) async {
```
nit:
```suggestion
                'and refocusing does not send "started typing" notice', (tester) async {
```

gnprice commented on test/widgets/compose_box_test.dart:
```diff
@@ -195,10 +206,163 @@ void main() {
+      checkTypingRequest(TypingOp.start, destinationNarrow);
+
+      connection.prepare(json: {});
+      // Move focus to the topic input.
+      await tester.tap(topicInputFinder);
+      await tester.pump(Duration.zero);
+      checkTypingRequest(TypingOp.stop, destinationNarrow);
```
Hmm, this actually seems a bit buggy. The user contemplates changing what topic they're sending to, and that causes us to decide they're not preparing a message after all?

If it's annoying to avoid this while getting the behavior we do want, then it's fine — *probably* the next thing the user is going to do is edit the topic, and then indeed we want to decide they're not sending to the old topic. But for the test, we should pick a scenario where the behavior is actually desirable.

I think the prime case for the "stop on unfocus" behavior is when the user touches the message list, or somewhere else outside the compose box, in order to unfocus the compose box and dismiss the keyboard.

---

… Oh and this also causes me to notice: on Android, once the focus is in the compose box it doesn't seem possible to get it out! At least, the cursor keeps blinking no matter what I do, short of navigating back from the whole page and then navigating into it again. This mostly doesn't affect the user — I think I hadn't previously noticed it in all this time using the app — but it does mean that this "stop on unfocus" behavior basically doesn't work on Android.

I think fixing that will be tough, though. I looked around and couldn't find an API for observing when the keyboard gets closed, or even checking whether it's open — the answers for that on the web are mostly just to look at the geometry of the window and try to guess.

So for this test, maybe just do this:
```dart
            FocusManager.instance.primaryFocus!.unfocus();
```

That's copied from the implementation in `ScrollView` of the semantics of `ScrollViewKeyboardDismissBehavior.onDrag`. Which is what we pass in `MessageList` to get the behavior we have on iOS where if you drag the message list it closes the keyboard. (The closing of the keyboard seems to come as a side effect of unfocusing the text input.)

gnprice commented on test/widgets/compose_box_test.dart:
```diff
@@ -195,10 +206,163 @@ void main() {
+    });
+
+    testWidgets('unfocusing app sends "typing stopped" notice', (tester) async {
+      Future<void> setAppLifeCycleState(AppLifecycleState state) async {
```
nit: match capitalization (i.e. match word boundaries)

gnprice commented on test/widgets/compose_box_test.dart:
```diff
@@ -195,10 +206,163 @@ void main() {
+        // While this state lives on [ServicesBinding], testWidgets resets it
+        // for us when the test ends:
+        //   https://github.com/flutter/flutter/blob/c78c166e3ecf963ca29ed503e710fd3c71eda5c9/packages/flutter_test/lib/src/binding.dart#L1189
+        final ByteData? message = const StringCodec().encodeMessage(state.toString());
+        await tester.binding.defaultBinaryMessenger
+          .handlePlatformMessage('flutter/lifecycle', message, (_) {});
```
Let's try to avoid reaching like this into the implementation details of how `didChangeAppLifecycleState` gets invoked.

See a similar case here:
https://github.com/zulip/zulip-flutter/pull/856#discussion_r1809705326

PIG208 replied:
That's pretty neat!

gnprice commented on test/widgets/compose_box_test.dart:
```diff
@@ -195,10 +206,163 @@ void main() {
+      connection.prepare(json: {});
+      await setAppLifeCycleState(AppLifecycleState.inactive);
+      checkTypingRequest(TypingOp.stop, narrow);
+      // Wait for the pending timer from the API request to end.
+      await tester.pump(Duration.zero);
+    });
```
Huh, why does this test case need this step when the other new test cases don't?

PIG208 replied:
When the request is triggered by an `enterText` call or a timeout, this is not needed. In the other cases, `await tester.pump` might have been called a bit earlier for different purposes, which also waits for the API request to end.

gnprice replied:
Ah I see, subtle. I think the point is that when timers get run, that leads to all timers getting run that are due at the ending time — which in particular means all fake API requests that are started at that time (and don't set a positive delay) get completed.

I think it might be less confusing if the last two lines are swapped, so that it ends the same kind of way as the other tests (and exactly the same way as the "navigating away" test, which is the most similar test case to this one):
```dart
      await tester.pump(Duration.zero);
      checkTypingRequest(TypingOp.stop, narrow);
```

That works, for the same reason the others work.

gnprice reviewed (commented):
Thanks for the revision! I've now reviewed the whole thing. Comments below.

Also replied in one subthread above: https://github.com/zulip/zulip-flutter/pull/897#discussion_r1818191351

PIG208 commented on lib/model/typing_status.dart:
```diff
@@ -84,3 +87,148 @@ class TypingStatus extends ChangeNotifier {
+    // We just started typing to this destination, so notify the server.
+    _currentDestination = destination;
+    _startOrExtendIdleTimer();
+    return _actuallyPingServer();
```
Hmm, I changed that [intentionally](https://github.com/zulip/zulip-flutter/pull/897#issuecomment-2439170063) to not `unawait` the future.

I see that removing `async` will make it unnecessary to `unawait` them, but it is also "discarding" the future somewhat non-obviously, which looks like one of the true-positives the discarded_futures rule would catch. Would it be clearer if we return them?

gnprice replied:
Returning a value when the function has type void effectively discards the value anyway — what `void` means is that the caller is supposed to not do anything with the value.

It'd be reasonable for the two call sites of `setTypingStatus` to use `unawaited` to make explicit that nothing is waiting for those requests.

PIG208 commented on test/widgets/compose_box_test.dart:
```diff
@@ -195,10 +206,163 @@ void main() {
+      final destinationNarrow = TopicNarrow(narrow.streamId, 'test topic');
+      await prepareComposeBox(tester, narrow: narrow, account: account);
+
+      await tester.enterText(topicInputFinder, destinationNarrow.topic);
+      // Remove an irrelevant topic request.
+      check(connection.takeRequests()).single
```
This confused me at first too. The response was prepared in `prepareComposeBox` whenever we call it with a `ChannelNarrow`. It would be cleaner if we just extract the `connection.prepare` out of the helper.

gnprice replied:
Hmm, I see.

It's tricky to extract that out of the helper, though, I think. That call happens in between the setup of the store and the setup of the widgets, both of which are handled by `prepareComposeBox`.

Can this `takeRequests()` and corresponding `check` be moved into `prepareComposeBox` instead? That way, they'd live in the same place as the `connection.prepare` for the same request. I think having those two sides live in the same place is the most important aspect for having this not be confusing.

PIG208 replied:
The method that makes the API request to topics does not get called [until the first update to the topic input occurs](https://github.com/zulip/zulip-flutter/blob/f86498e21f6330d0202699f1130293878fbb6e0c/lib/widgets/autocomplete.dart#L43-L45). So moving `connection.prepare` right before `enterText` should work fine.

gnprice replied:
That seems like it would work but also make the call sites more complex.

The fact that the topic input causes these requests is beside the point for most of these tests — it's not related to what the tests are about. So if they can avoid having to deal with it, that helps keep the tests focused.

PIG208 commented:
This hasn't addressed everything yet. I will push another update later today. The current revision shows the general approach.

I think the later changes for sending typing notices for scrolling list of autocompletion items and interacting with compose buttons might be better done in a separate PR that fixes the issue, as this thread has become longer.

gnprice commented:
Sounds good! I agree a follow-up PR is good for those.

PIG208 commented:
Created #1033 as a follow-up, and left the app lifecycle change here with the main implementation. Thanks for the review! This is ready now.

gnprice commented on test/widgets/compose_box_test.dart:
```diff
@@ -41,16 +56,26 @@ void main() {
       child: ComposeBox(controllerKey: controllerKey, narrow: narrow)));
     await tester.pumpAndSettle();
 
+    if (topic != null) {
+      // The topic input is currently only available to ChannelNarrow.
+      narrow as ChannelNarrow;
+      connection.prepare(body:
+        jsonEncode(GetStreamTopicsResult(topics: [eg.getStreamTopicsEntry()]).toJson()));
+      await tester.enterText(topicInputFinder, topic);
+      check(connection.takeRequests()).single
```
Hmm, I see. So the test cases that cause this request to happen are only the handful that do something to the topic input? I guess you were pointing at that in the previous comment https://github.com/zulip/zulip-flutter/pull/897#discussion_r1823699207 , but I didn't see the implication.

gnprice replied:
Given that it's only two existing test cases (plus a couple of new ones) that turn out to actually need this functionality, and that it can just as well be done after the rest of the setup is complete instead of mixed into the middle of the other setup, it'd be best to do it separately. That way this prep helper doesn't grow into an omnibus do-everything, and it stays more straightforward to see what each of the tests is doing.

For the topic input, because there are these three steps involved — prepare request, enter text, consume request — it'd make sense to have a helper for those. Then the test cases that need it can call that helper, after calling `prepareComposeBox`.

For the content input, the step is trivial, so test cases should just do that step directly to keep things straightforward.

gnprice commented on test/widgets/compose_box_test.dart:
```diff
@@ -264,14 +455,15 @@ void main() {
         // the file is uploading, not a pre-existing reason.)
-        composeBoxController.topicController!.value = const TextEditingValue(text: 'some topic');
-        composeBoxController.contentController.value = const TextEditingValue(text: 'see image: ');
-        await tester.pump();
```
This comment no longer makes sense — it was here to explain why we're setting the topic and content.

gnprice commented on test/widgets/compose_box_test.dart:
```diff
@@ -195,15 +220,181 @@ void main() {
+      checkTypingRequest(TypingOp.stop, narrow);
+    });
+
+    testWidgets('for content input, unfocusing sends a "typing stopped" notice '
+                'and refocusing sends a "typing started" notice', (tester) async {
```
> refocusing sends a "typing started" notice

Is this desirable? It seems undesirable to me.

I think I see why it happens in the current implementation. It's not a big deal and we don't have to fix it before merge. But as I said in the previous thread:
https://github.com/zulip/zulip-flutter/pull/897#discussion_r1818188643
the test should focus on testing behaviors we actually want. Moreover if we do have a test that checks behavior we don't want, it should be very clearly labeled as such — otherwise, the test becomes an extra barrier in the way of us fixing the behavior.

gnprice commented on lib/widgets/compose_box.dart:
```diff
@@ -269,19 +269,101 @@ class ComposeContentController extends ComposeController<ContentValidationError>
+      // the user started typing, so do nothing.
+      return;
+    }
+    // Losing focus usually indicates that the user has navigated away or
+    // clicked on other UI elements.  On Android, this does not get triggered
+    // unless the user navigates away from the page or focuses on another input.
+    // See: https://github.com/zulip/zulip-flutter/pull/897#discussion_r1818188643
+    final store = PerAccountStoreWidget.of(context);
+    store.typingNotifier.stoppedComposing();
```
```suggestion
    final store = PerAccountStoreWidget.of(context);
    store.typingNotifier.stoppedComposing();
```

The "On Android" part doesn't seem relevant to understanding this code; it's not part of why we wrote this code the way we did. (If hiding the keyboard did unfocus the input, we'd be glad and would still want to say the user has stopped typing.) So it can be left out.

I'm also not sure that comment is accurate — it tries to give a complete list of the things that will cause unfocus, and I'm not sure there aren't other things we haven't thought of that also do so. (What about opening the bottom sheet; does that unfocus the input? I suspect it does.)

Then the "Losing focus" sentence seems like it's just explaining the meaning of focus.

gnprice reviewed (commented):
Thanks for the revision! Comments below.

PIG208 commented on lib/widgets/compose_box.dart:
```diff
@@ -370,6 +452,8 @@ class _StreamContentInputState extends State<_StreamContentInput> {
     return _ContentInput(
       narrow: widget.narrow,
+      destination: TopicNarrow(
+        widget.narrow.streamId, widget.topicController.textNormalized),
```
Thanks for pointing that out. It was unintentional that I used `widget.topicController.textNormalized` here.

PIG208 commented:
Thanks for the review! I have updated the PR addressing the comments. Note that I mentioned a minor issue in [this discussion on `composing` and `_focusChanged`](https://chat.zulip.org/#narrow/channel/243-mobile-team/topic/PR.20.23F897.3A.20TextEditingValue.20and.20keyboard.20status), which is not resolved in this update; my current plan to do so in a follow-up PR.

gnprice commented on test/widgets/compose_box_test.dart:
```diff
@@ -54,6 +57,21 @@ void main() {
     return controllerKey;
   }
 
+  Future<void> checkEnterTopic(WidgetTester tester, {
```
nit:
```suggestion
  Future<void> enterTopic(WidgetTester tester, {
```

The purpose of this function is to enter a topic, as setup for whatever the calling test is going to do next, rather than to check anything.

(There is a check but it's there because we need to consume the request in order to leave things clean for the rest of the test; the check is useful for validating that it was a get-stream-topics request, so that if something unexpected happens instead of that, we aren't just confusing things by consuming a stray request.)

So because the check is incidental and not the point of the function, the name should reflect what the function is about, to let the reader at a call site focus their attention on the checks that the test is meant to focus on.

gnprice reviewed (commented):
Thanks for the revision! This all looks good modulo one nit below, and one commit message that looks like it's meant to be updated before merge:
b0f8c352f fix link; binding: Add stopwatch.

So please go ahead and make that update along with fixing the nit, and then I think it'll be time to merge this.

PIG208 commented:
Thanks for the review! Updated the PR.

gnprice commented:
Thanks! Looks good; merging.
